### PR TITLE
feat: Implement low-level wasm i32/i64/f32/f64 libraries

### PR DESCRIPTION
**Pull request by ospencer**
## Summary
This PR adds a new way for contributors to write standard libraries. It adds four new Grain types—`WasmI32`, `WasmI64`, `WasmF32`, and `WasmF64`—and four new libraries to operate on them. There is one additional library for some memory operations.

## Motivation
Sometimes when we write standard library functions we need to perform low-level WebAssembly actions. An example of this might be converting the bytes of a `String` into a `Char` array. Typically, this is done by adding a new compiler `primitive` or by referencing a function written outside of Grain via the `foreign` syntax. Often, we only never need just a few instructions, so having to leave the context of our Grain file feels pretty cumbersome. Additionally, when we're not working within a Grain file, we don't have access to Grain constructs that might be helpful. This PR allows us to directly use WebAssembly instructions from our Grain code, potentially like so:
```grain
import WasmI32 from "unsafe/wasmi32"

@disableGC
let isNegative = (x) => {
  WasmI32.eqz(WasmI32.clz(x))
}
```
> _This could be more efficiently implemented in quite a few ways, but please enjoy my contrived example._

In the future, this should allow us to perform better optimizations and potentially perform static linking against our standard libraries.

### Type representations
The main reason why this PR is so large is that up until this point, all user-usable Grain values were WebAssembly `i32`s. Most of the code in this PR is allowing types to have representations other than `i32`. Previously, type declarations had a field called `type_immediate` which indicated if the value was a pointer to a value on the heap or if the value exists on the stack. I've replaced this with `type_allocation`, which has a value of `HeapAllocated` or `StackAllocated(wasm_type)`. This information is used to tell Grain what WebAssembly type to use for locals, etc. This information could also be used to optimize calls to the GC—only `HeapAllocated` values need a call to `incRef` or `decRef`.

### Function compilation
Generated WebAssembly functions now are properly typed—if it accepts types with representations other than `i32` or returns a type with a representation other than `i32`, it'll have those types. (This wasn't necessary before as all Grain values were `i32`.) This also has an impact on the FFI—Grain can now use external functions that accept or return types other than `i32`.

### Optimization
Included with this change is an optimization pass that replaces calls to any of the wasm instruction functions with the actual instruction. However, since each instruction is a real Grain function, they can also be used as first-class functions. The only downside is that if used this way, it will be an actual function call instead of a single WebAssembly instruction.

### Literals
Each of the four new types comes with its own literal syntax. `1n`, `1N`, `1.2w`, and `1.2W` for types `WasmI32`, `WasmI64`, `WasmF32`, and `WasmF64` respectively. The `n` syntax comes from OCaml, where the `n` stands for "native" (though here it's used slightly differently). I feel that the added syntax is justified because of the sheer number of times a constant value must be provided to an operation (and our `store` operations _always_ require an offset, which must be an immediate). The syntax doesn't clash with anything else, and can really only be "accidentally" discovered through typos.

### Attributes
As these types don't do well with the GC (see the section on Limitations), there is now support for attributes on bindings. The only one currently implemented is the `@disableGC` attribute, which can be used to tell the compiler to not emit any GC instructions when compiling the function. We can add more attributes to do other things in the future.

## Limitations
There are some limitations to using these types (which is one of the reasons why these libraries are unsafe).

- These types do not play well with the GC, as none of the values are tagged. As such, any use of these types must be done in an unmanaged memory context, either by disabling the GC globally or using the `@disableGC` attribute. In such a context, any explicit or implicit allocations must be manually `free`d or `decRef`'d. This can potentially change (see earlier section on type representations).
- These types do not work with `let mut` because of `let mut`'s underlying implementation. This may change in the future with some (already planned) optimizations.
- These types cannot currently be used in globals, though this is not impossible to fix.
- Functions with "any" types, e.g. `print` and `(==)` don't know how to handle these types, and cannot be used with them. However, each new module provides a `print` implementation for debugging and `eq` from the module can be used instead of `==`.

### `import WasmI32 from "unsafe/wasmi32"`
```grain
// WebAssembly Memory Instructions
load : (WasmI32, WasmI32) -> WasmI32
load8S : (WasmI32, WasmI32) -> WasmI32
load8U : (WasmI32, WasmI32) -> WasmI32
load16S : (WasmI32, WasmI32) -> WasmI32
load16U : (WasmI32, WasmI32) -> WasmI32
store : (WasmI32, WasmI32, WasmI32) -> Void
store8 : (WasmI32, WasmI32, WasmI32) -> Void
store16 : (WasmI32, WasmI32, WasmI32) -> Void

// WebAssembly Unary Instructions
clz : WasmI32 -> WasmI32
ctz : WasmI32 -> WasmI32
popcnt : WasmI32 -> WasmI32
eqz : WasmI32 -> Bool

// WebAssembly Binary Instructions
add : (WasmI32, WasmI32) -> WasmI32
sub : (WasmI32, WasmI32) -> WasmI32
mul : (WasmI32, WasmI32) -> WasmI32
divS : (WasmI32, WasmI32) -> WasmI32
divU : (WasmI32, WasmI32) -> WasmI32
remS : (WasmI32, WasmI32) -> WasmI32
remU : (WasmI32, WasmI32) -> WasmI32
and : (WasmI32, WasmI32) -> WasmI32
or : (WasmI32, WasmI32) -> WasmI32
xor : (WasmI32, WasmI32) -> WasmI32
shl : (WasmI32, WasmI32) -> WasmI32
shrS : (WasmI32, WasmI32) -> WasmI32
shrU : (WasmI32, WasmI32) -> WasmI32
rotl : (WasmI32, WasmI32) -> WasmI32
rotr : (WasmI32, WasmI32) -> WasmI32
eq : (WasmI32, WasmI32) -> Bool
ne : (WasmI32, WasmI32) -> Bool
ltS : (WasmI32, WasmI32) -> Bool
ltU : (WasmI32, WasmI32) -> Bool
leS : (WasmI32, WasmI32) -> Bool
leU : (WasmI32, WasmI32) -> Bool
gtS : (WasmI32, WasmI32) -> Bool
gtU : (WasmI32, WasmI32) -> Bool
geS : (WasmI32, WasmI32) -> Bool
geU : (WasmI32, WasmI32) -> Bool

// WebAssembly Conversion Instructions
wrapI64 : WasmI64 -> WasmI32
truncF32S : WasmF32 -> WasmI32
truncF32U : WasmF32 -> WasmI32
truncF64S : WasmF64 -> WasmI32
truncF64U : WasmF64 -> WasmI32
reinterpretF32 : WasmF32 -> WasmI32
extendS8 : WasmI32 -> WasmI32
extendS16 : WasmI32 -> WasmI32

// Grain Conversions
ofGrain : a -> WasmI32
toGrain : WasmI32 -> a
toInt32 : WasmI32 -> Int32
ofInt32 : Int32 -> WasmI32

// Utilities
print : WasmI32 -> Void
```

### `import WasmI64 from "unsafe/wasmi64"`
```grain
// WebAssembly Memory Instructions
load : (WasmI32, WasmI32) -> WasmI64
load8S : (WasmI32, WasmI32) -> WasmI64
load8U : (WasmI32, WasmI32) -> WasmI64
load16S : (WasmI32, WasmI32) -> WasmI64
load16U : (WasmI32, WasmI32) -> WasmI64
load32S : (WasmI32, WasmI32) -> WasmI64
load32U : (WasmI32, WasmI32) -> WasmI64
store : (WasmI32, WasmI64, WasmI32) -> Void
store8 : (WasmI32, WasmI64, WasmI32) -> Void
store16 : (WasmI32, WasmI64, WasmI32) -> Void
store32 : (WasmI32, WasmI64, WasmI32) -> Void

// WebAssembly Unary Instructions
clz : WasmI64 -> WasmI64
ctz : WasmI64 -> WasmI64
popcnt : WasmI64 -> WasmI64
eqz : WasmI64 -> Bool

// WebAssembly Binary Instructions
add : (WasmI64, WasmI64) -> WasmI64
sub : (WasmI64, WasmI64) -> WasmI64
mul : (WasmI64, WasmI64) -> WasmI64
divS : (WasmI64, WasmI64) -> WasmI64
divU : (WasmI64, WasmI64) -> WasmI64
remS : (WasmI64, WasmI64) -> WasmI64
remU : (WasmI64, WasmI64) -> WasmI64
and : (WasmI64, WasmI64) -> WasmI64
or : (WasmI64, WasmI64) -> WasmI64
xor : (WasmI64, WasmI64) -> WasmI64
shl : (WasmI64, WasmI64) -> WasmI64
shrU : (WasmI64, WasmI64) -> WasmI64
shrS : (WasmI64, WasmI64) -> WasmI64
rotl : (WasmI64, WasmI64) -> WasmI64
rotr : (WasmI64, WasmI64) -> WasmI64
eq : (WasmI64, WasmI64) -> Bool
ne : (WasmI64, WasmI64) -> Bool
ltS : (WasmI64, WasmI64) -> Bool
ltU : (WasmI64, WasmI64) -> Bool
leS : (WasmI64, WasmI64) -> Bool
leU : (WasmI64, WasmI64) -> Bool
gtS : (WasmI64, WasmI64) -> Bool
gtU : (WasmI64, WasmI64) -> Bool
geS : (WasmI64, WasmI64) -> Bool
geU : (WasmI64, WasmI64) -> Bool

// WebAssembly Conversion instructions
extendI32S : WasmI32 -> WasmI64
extendI32U : WasmI32 -> WasmI64
truncF32S : WasmF32 -> WasmI64
truncF32U : WasmF32 -> WasmI64
truncF64S : WasmF64 -> WasmI64
truncF64U : WasmF64 -> WasmI64
reinterpretF64 : WasmF64 -> WasmI64
extendS8 : WasmI64 -> WasmI64
extendS16 : WasmI64 -> WasmI64
extendS32 : WasmI64 -> WasmI64

// Grain Conversions
toInt64 : WasmI64 -> Int64
ofInt64 : Int64 -> WasmI64

// Utilities
print : WasmI64 -> Void
```

### `import WasmF32 from "unsafe/wasmf32"`
```grain
// WebAssembly Memory Instructions
load : (WasmI32, WasmI32) -> WasmF32
store : (WasmI32, WasmF32, WasmI32) -> Void

// WebAssembly Unary Instructions
neg : WasmF32 -> WasmF32
abs : WasmF32 -> WasmF32
ceil : WasmF32 -> WasmF32
floor : WasmF32 -> WasmF32
trunc : WasmF32 -> WasmF32
nearest : WasmF32 -> WasmF32
sqrt : WasmF32 -> WasmF32

// WebAssembly Binary Instructions
add : (WasmF32, WasmF32) -> WasmF32
sub : (WasmF32, WasmF32) -> WasmF32
mul : (WasmF32, WasmF32) -> WasmF32
div : (WasmF32, WasmF32) -> WasmF32
copySign : (WasmF32, WasmF32) -> WasmF32
min : (WasmF32, WasmF32) -> WasmF32
max : (WasmF32, WasmF32) -> WasmF32
eq : (WasmF32, WasmF32) -> Bool
ne : (WasmF32, WasmF32) -> Bool
lt : (WasmF32, WasmF32) -> Bool
le : (WasmF32, WasmF32) -> Bool
gt : (WasmF32, WasmF32) -> Bool
ge : (WasmF32, WasmF32) -> Bool

// WebAssembly Conversion instructions
reinterpretI32 : WasmI32 -> WasmF32
convertI32S : WasmI32 -> WasmF32
convertI32U : WasmI32 -> WasmF32
convertI64S : WasmI64 -> WasmF32
convertI64U : WasmI64 -> WasmF32
demoteF64 : WasmF64 -> WasmF32

// Grain Conversions
toFloat32 : WasmF32 -> Float32
ofFloat32 : Float32 -> WasmF32

// Utilities
print : WasmF32 -> Void
```

### `import WasmF64 from "unsafe/wasmf64"`
```grain
// WebAssembly Memory Instructions
load : (WasmI32, WasmI32) -> WasmF64
store : (WasmI32, WasmF64, WasmI32) -> Void

// WebAssembly Unary Instructions
neg : WasmF64 -> WasmF64
abs : WasmF64 -> WasmF64
ceil : WasmF64 -> WasmF64
floor : WasmF64 -> WasmF64
trunc : WasmF64 -> WasmF64
nearest : WasmF64 -> WasmF64
sqrt : WasmF64 -> WasmF64

// WebAssembly Binary Instructions
add : (WasmF64, WasmF64) -> WasmF64
sub : (WasmF64, WasmF64) -> WasmF64
mul : (WasmF64, WasmF64) -> WasmF64
div : (WasmF64, WasmF64) -> WasmF64
copySign : (WasmF64, WasmF64) -> WasmF64
min : (WasmF64, WasmF64) -> WasmF64
max : (WasmF64, WasmF64) -> WasmF64
eq : (WasmF64, WasmF64) -> Bool
ne : (WasmF64, WasmF64) -> Bool
lt : (WasmF64, WasmF64) -> Bool
le : (WasmF64, WasmF64) -> Bool
gt : (WasmF64, WasmF64) -> Bool
ge : (WasmF64, WasmF64) -> Bool

// WebAssembly Conversion instructions
reinterpretI64 : WasmI64 -> WasmF64
convertI32S : WasmI32 -> WasmF64
convertI32U : WasmI32 -> WasmF64
convertI64S : WasmI64 -> WasmF64
convertI64U : WasmI64 -> WasmF64
promoteF32 : WasmF32 -> WasmF64

// Grain Conversions
toFloat64 : WasmF64 -> Float64
ofFloat64 : Float64 -> WasmF64

// Utilities
print : WasmF64 -> Void
```

### `import Memory from "unsafe/memory"`

```grain
malloc : WasmI32 -> WasmI32
free : WasmI32 -> Void
incRef : WasmI32 -> WasmI32
decRef : WasmI32 -> WasmI32

copy : (WasmI32, WasmI32, WasmI32) -> Void
fill : (WasmI32, WasmI32, WasmI32) -> Void
```

### CODE REVIEW

**Comment by tomByrer** [comment]
One of the main reasons I'd use WASM is for DSP.  It is very helpful to have other number types other than i32, but how much of a hassle is the lack of Garbage Collection ?  (I've never programmed a manual GC language; why I avoided C).

**Comment by ospencer** [comment]
Hey @tomByrer! This change is a bit of an "under-the-hood" change that makes things easier for people who contribute to the language and potentially library authors in the future, but it's not meant for everyday programming. Grain already has support for 32/64-bit ints/floats that work with the GC. Here's the [Float32 library](https://grain-lang.org/docs/stdlib/float32), for example.

The reason why using these new libraries would result in having to manage your own memory is because you're not allowing the compiler to generate the extra code necessary to correctly do garbage collection. You can think of the pre-existing number libraries as all of the same operations, but with all of the safeguards that you'd want from a high-level programming language.

Hope that makes sense!

File: compiler/src/parsing/ast_helper.re [hunk 1]
```diff
@@ -177,69 +181,115 @@ module Pat = {
...
       switch (loc) {
       | None => default_loc_src^()
       | Some(l) => l
       };
-    {pexp_desc: d, pexp_loc: loc};
+    let attributes =
+      switch (attributes) {
+      | None => []
+      | Some(l) => l
+      };
```
Comment (phated) [inline compiler/src/parsing/ast_helper.re:194]: `Option.default`?

Comment (phated) [reply compiler/src/parsing/ast_helper.re:194]: I'm fine with this being a follow up PR to change all these in this file.

File: compiler/src/parsing/parsetree.re [hunk 2]
```diff
@@ -167,7 +311,29 @@ type prim1 =
   | Int32ToNumber
   | Float64ToNumber
   | Float32ToNumber
-  | Int64Lnot;
+  | Int64Lnot
+  | WasmOfGrain
```
Comment (phated) [inline compiler/src/parsing/parsetree.re:315]: `WasmFromGrain`?

**Review by phated — COMMENTED** [review]
LGTM! We went over this together on a call ❤️

**Review by phated — APPROVED** [review]
Love the changes! ❤️ 🎉

File: compiler/src/codegen/mashtree.re [hunk 3]
```diff
@@ -40,7 +184,29 @@ type prim1 =
     | Int32ToNumber
     | Float64ToNumber
     | Float32ToNumber
-    | Int64Lnot;
+    | Int64Lnot
+    | WasmOfGrain
```
[the remark this answers is not in the document]
Comment (phated) [reply compiler/src/codegen/mashtree.re:192]: @ospencer Did we keep these variants `WasmOfGrain` still? I know it's an internal thing, but weird that they are different than the external facing naming.

[the remark this answers is not in the document]
Comment (ospencer) [reply compiler/src/codegen/mashtree.re]: Fixed!

**Pull request merged**